client/my-sites/add-ons/utils/get-feature-from-product.ts

```diff
@@ -0,0 +1,15 @@
+import {
+	WPCOM_FEATURES_UNLIMITED_THEMES,
+	WPCOM_FEATURES_PREMIUM_THEMES,
+} from '@automattic/calypso-products';
+
+/** Sometimes features slugs don't map directly to product slugs. */
+const getFeatureFromProduct = ( productSlug: string ) => {
+	if ( WPCOM_FEATURES_UNLIMITED_THEMES === productSlug ) {
+		return WPCOM_FEATURES_PREMIUM_THEMES;
+	}
+
+	return productSlug;
+};
+
+export default getFeatureFromProduct;
```

Comment:
Could this function be removed after replacing mentions of `WPCOM_FEATURES_UNLIMITED_THEMES` with `WPCOM_FEATURES_PREMIUM_THEMES`?

Reply:
I'm not sure. Right now, one is a product, the other is a feature (not returned as product or have a checkout handle) and we need both contexts for this page. It's confusing - not sure feel excited either about renaming just this case into `WPCOM_PRODUCT_...` without establishing a pattern. Might be a good direction though if done for all.

Any thoughts? Fwiw there may be more add-ons coming that don't map to features (like storage).

Reply:
Or are you referring to "backend" mentions?

(I'd need to come back with a clear mind and understand a little better why we end up with two) 🤷‍♂️ 

@dzver

Reply:
There is [prior art for WP.com products in Calypso](https://github.com/Automattic/wp-calypso/blob/c53ca8cadaee2e1f873803daa6336672c3058cc9/packages/calypso-products/src/constants/wpcom.ts#L4-L6), so whenever you want to check if a site has purchased the product, that should probably be used. From what I can tell, [this should be the only place](https://github.com/Automattic/wp-calypso/pull/64289/files#diff-4387dc1fa32a1f09874042eebe9fb71ebda9089594a3507b1d620d76af89af40R51) where you'd need the product slug.

`siteHasFeature` determines if the site has _any_ product that gives it that feature. That mapping is defined in WPCOM_Features in the wpcom repository and should include the three add-on products mapped to these three features. 

> Fwiw there may be more add-ons coming that don't map to features (like storage)
Yeah, storage will be an interesting one. I assume every add-on provides a feature of some kind, so if it doesn't exist in WPCOM_Features yet it should be created there to be used with `siteHasFeature`.

Reply:
Thanks @obenland . Glad to see that prior art! So will gladly add some slugs there :-)

My understanding of all these is fairly new (days' worth). AFAIU from the flow here, `product !== feature` (at least not necessarily), and it seems somewhat coincidental that plan features have the same slugs as defined products. To put it differently, I don't think storage should behave any differently than the rest. 

Most of this comes from checking the availability of add-ons for purchase, and providing an appropriate message for each context:

1. If add-on is added to cart (in Signup) -> _"Included in cart"_
2. If add-on is included as a feature in site plan -> _"Included in your plan"_
3. If user purchased the add-on directly -> _"Purchased"_

(given, differentiating on 3 above came during implementation and chat with @dzver, but it might just be easier to avoid and save us the trouble)

So, given all three contexts, if `siteHasFeature` looks beyond checking some pre-association of plans to features (e.g. like [getAllFeaturesForPlans](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-products/src/main.ts#L163) does with the frontend mappings - if I understand), then it may not fit this use case. In other words, if it also includes products purchased on a given site, then we won't be able to differentiate between 2 & 3 above. 

I was hoping `isSiteFeature` was a convenient way of checking the return value of the `/features` endpoint, hence avoiding the trouble of associating plans to features on the frontend. but something about the name was indeed raising concerns that it's probably more than that (which I ignored) 🤷‍♂️

Reply:
> So, given all three contexts, if siteHasFeature looks beyond checking some pre-association of plans to features (e.g. like [getAllFeaturesForPlans](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-products/src/main.ts#L163) does with the frontend mappings - if I understand), then it may not fit this use case. In other words, if it also includes products purchased on a given site, then we won't be able to differentiate between 2 & 3 above.

> I was hoping siteHasFeature was a convenient way of checking the return value of the /features endpoint, hence avoiding the trouble of associating plans to features on the frontend. but something about the name was indeed raising concerns that it's probably more than that (which I ignored) 🤷‍♂️

Well, actually - I am wrong. Nothing wrong with it doing more than that right? We do early return on a "purchases" check and handle the "features" check second - as we do already. 😁 

Sounds like a false alarm then! 🙇

Reply:
Yeah, I think you're running into the most complicated aspects of Calypso's patch work of plans and features mapping, where `siteHasFeature` is based on the `/features` endpoint but `getAllFeaturesForPlans` is based on the hard-coded [plans-list.jsx](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-products/src/plans-list.tsx).

For the purpose of this particular piece of code, yes, it _is_ coincidental that some features have the same slugs as defined products so they should be looked at separately and not derived from one-another.

Reply:
Huh, I didn't get a notification about this comment until after I wrote [that one](https://github.com/Automattic/wp-calypso/pull/64289#discussion_r907603245).

I think my preference would be to avoid a 1:1 relationship between products and features (because they're really 1:many) and hard-code the product and feature to check for.

Reply:
> Huh, I didn't get a notification about this comment until after I wrote https://github.com/Automattic/wp-calypso/pull/64289#discussion_r907603245.

Heh I was not the only one then! Notifications were not coming through from GH yesterday. Replied [here](https://github.com/Automattic/wp-calypso/pull/64289#discussion_r908280024).